### PR TITLE
[BUGFIX beta] Fix no-jQuery EventDispatcher to prevent events bubbling up when event.stopPropagation() was called

### LINKED ISSUE #17840: event.stopPropagation() no longer stops propagation after removing jQuery
Hi!

I've built and am maintaining the [ember-drag-sort](https://github.com/kaliber5/ember-drag-sort) addon which happens to be the only drag-n-drop addon in the Ember ecosystem with support for nested lists.

It's been working fine with jQuery. Demo: https://kaliber5.github.io/ember-drag-sort/

I've tried upgrading to Ember 3.8 without jQuery and discovered that `event.stopPropagation()` no longer stops propagation. As a result, events on child lists get overridden by events on parent lists, and reordering nested lists no longer works.

I've pushed the `gen-1` branch which is based on no-jQuery Ember 3.8. Here are some of problematic spots:

* https://github.com/kaliber5/ember-drag-sort/blob/gen-1/addon/components/drag-sort-list.js#L103
* https://github.com/kaliber5/ember-drag-sort/blob/gen-1/addon/components/drag-sort-item.js#L152

(Note that in the latter spot I tried adding `event.preventDefault()` and `return false` which didn't help and might have made things even worse, so it might make sense removing those when you try.)

This looks like a regression in Ember to me, but I can't confirm it myself. Please help!

### PR DESCRIPTION
**Pull request by simonihmig**
Fixes #17840

### CODE REVIEW

**Review by chancancode — APPROVED** [review]
:clap:

**Comment by chancancode** [comment]
Does return false work?

File: packages/@ember/-internals/views/lib/system/event_dispatcher.js [hunk 1]
```diff
@@ -360,6 +360,8 @@ export default EmberObject.extend({
                 event.preventDefault();
                 event.stopPropagation();
                 break;
+              } else if (event.cancelBubble === true) {
+                break;
```
Comment (chancancode) [inline packages/@ember/-internals/views/lib/system/event_dispatcher.js]: This is probably not the correct nesting for this. I believe we want to check the flag after the event handler has run (currently the check runs when if there is no event handler). Further, we probably need the same treatment for the action branch too. So I think you probably want to do it before setting the target to the parent.

I feel like this should have failed the test tho?

**Review by chancancode — CHANGES_REQUESTED** [review]

File: packages/@ember/-internals/views/lib/system/event_dispatcher.js (hunk 1 above)
Comment (chancancode) [reply packages/@ember/-internals/views/lib/system/event_dispatcher.js]: Also, we should make sure to mirror the jquery/dom behavior that stop propagation doesn’t prevent handlers on the same element from firing

Comment (chancancode) [reply packages/@ember/-internals/views/lib/system/event_dispatcher.js]: On that note, we probably have another bug here. It seems like the code assumes you can either have a component element handler, or action handlers, but not both. I believe with the possibility of forwarding action modifiers through splattributes, this is no longer a correct assumption.

Comment (simonihmig) [reply packages/@ember/-internals/views/lib/system/event_dispatcher.js]: > I believe we want to check the flag after the event handler has run

In what scenario would that not be the case? `viewHandler(target, event)` will always be called before our `cancelBubble` check, right? 

> Further, we probably need the same treatment for the action branch too.

I tried to add a (failing) test for that, until I realized that the action handler will not receive a reference to the event, so cannot call `stopPropagation()`, or do I miss something?

> stop propagation doesn’t prevent handlers on the same element from firing

How would there be two different handlers (component event handlers) on the same element? You mean the case you mentioned in your following comment? 👇

> It seems like the code assumes you can either have a component element handler, or action handlers, but not both. I believe with the possibility of forwarding action modifiers through splattributes, this is no longer a correct assumption.

Hm, I see... 🤔Good catch! But we should tackle this in a separate PR I think?

**Comment by simonihmig** [comment]
> Does return false work?

Yes, I think so! This was already  covered in a test: https://github.com/emberjs/ember.js/blob/561759e4d3cd2f978deb057e4e1f52d03098f40d/packages/%40ember/-internals/glimmer/tests/integration/event-dispatcher-test.js#L109

File: packages/@ember/-internals/views/lib/system/event_dispatcher.js (hunk 1 above)
Comment (chancancode) [reply packages/@ember/-internals/views/lib/system/event_dispatcher.js]: > `viewHandler(target, event)` will always be called before our `cancelBubble` check, right?

You're right!


> action handler will not receive a reference to the event

Don't we pass it [here](https://github.com/emberjs/ember.js/blob/561759e4d3cd2f978deb057e4e1f52d03098f40d/packages/%40ember/-internals/views/lib/system/event_dispatcher.js#L292)?

> How would there be two different handlers (component event handlers) on the same element? You mean the case you mentioned in your following comment? 👇

Yep that's one case I could think of. I think you could also do `<div {{action "foo"}} {{action "bar"}} />`? Does that work? (Looking at the code, I'm not sure where this is handled.)

> But we should tackle this in a separate PR I think?

Seems fine to me!

Comment (simonihmig) [reply packages/@ember/-internals/views/lib/system/event_dispatcher.js]: > Don't we pass it here?

Yes, but that goes to https://github.com/emberjs/ember.js/blob/83263728b116fac6d4e8fcc798babf555a62ac2b/packages/%40ember/-internals/glimmer/lib/modifiers/action.ts#L125-L181, and there the event is not passed to the action handler. I believe this is how the action helper API works: the (element space(!), not closure) action helper passes only the custom args, not the event itself. That's why we have things like `bubbles=false`.

>  I think you could also do <div {{action "foo"}} {{action "bar"}} />? Does that work? (Looking at the code, I'm not sure where this is handled.)

We have that loop here: https://github.com/emberjs/ember.js/blob/561759e4d3cd2f978deb057e4e1f52d03098f40d/packages/%40ember/-internals/views/lib/system/event_dispatcher.js#L288

And a related test here, though for different events: https://github.com/emberjs/ember.js/blob/561759e4d3cd2f978deb057e4e1f52d03098f40d/packages/%40ember/-internals/glimmer/tests/integration/helpers/element-action-test.js#L1323-L1360

I tried to add a similar test with the same event type, but that fails, as the first handled action handler returns early here: https://github.com/emberjs/ember.js/blob/561759e4d3cd2f978deb057e4e1f52d03098f40d/packages/%40ember/-internals/views/lib/system/event_dispatcher.js#L292

So this seems to be not supported at this point. I guess it should be possible, by always looping over all action handlers, and return `false` if at least one returned `false`!?

Comment (simonihmig) [reply packages/@ember/-internals/views/lib/system/event_dispatcher.js]: Oh, wait a second. Actually that test with multiple `{{action}}` *does*  pass w/ jQuery, so it seems it should actually work, but doesn't in no-jQuery mode! 

I'll bring up another PR to fix this specific issue. As for the issue in this PR, I *think* this PR should be ok as it is?

Comment (simonihmig) [reply packages/@ember/-internals/views/lib/system/event_dispatcher.js]: Here we go: #17876

**Pull request merged**